### PR TITLE
Get instance for extra processes

### PR DESCRIPTION
**Pull request by stephendb**
- Adds a multiprocess example
- Adds a new call, SampleSensor which takes a list of sensors and only samples those sensors as opposed to all sensors on server.
- Adds connectivity to the getInstance call and removes configuration requirement for processes that are only going to send commands and are not configuring.

### CODE REVIEW

File: examples/cpp/lane_follower/fixed_step_lidar.cpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    };
+
+    // start our main perception planning and control loop
+    std::cout << "Sampling sensor loop" << std::endl;
+    while(true)
+    {
+        // step the simulation one simulation frame
+        ClosedLoopStepCommandConfig stepCommand;
+        stepCommand.time_step = 0.016;
+        sim0.sendCommand(ApiMessage(1234, ClosedLoopStepCommand_ID, true, stepCommand.dump()));
```
Comment (bradleynull) [inline examples/cpp/lane_follower/fixed_step_lidar.cpp:202]: You can use the `ClosedLoopStepCommandConfig` here along with its `.message()` call.

File: examples/cpp/multi_process_example/simulator_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    Simulator& sim0 = Simulator::getInstance(config, server0_ip, server_port);
+
+    if(!sim0.configure()){
+        return -1;
+    }
+
+    while(true)
+    {
+    }
```
Comment (bradleynull) [inline examples/cpp/multi_process_example/simulator_controller.cpp:31]: Should we put a no-op here like `std::chrono::sleep_for(std::chrono::nanoseconds(1)`. It seems like this my tie up a core at 100%.

File: monodrive/core/src/Simulator.cpp [hunk 3]
```diff
@@ -26,16 +26,25 @@ Simulator::Simulator(const Configuration& config, const std::string& serverIp, c
...
 Simulator& Simulator::getInstance(const std::string& serverIp, const short& serverPort)
 {
-		std::lock_guard<std::mutex> simLock(_mutex);
-		const std::string simKey = serverIp + ":" + std::to_string(serverPort);
-		return *simMap[simKey];
+	std::lock_guard<std::mutex> simLock(_mutex);
+	const std::string simKey = serverIp + ":" + std::to_string(serverPort);
+	if (simMap[simKey] == nullptr) {
+		simMap[simKey] = new Simulator(serverIp, serverPort);
+		std::cout << "created new simulator:" << simKey << std::endl;
```
Comment (bradleynull) [inline monodrive/core/src/Simulator.cpp:44]: Remove debug output.

File: monodrive/core/src/Simulator.cpp [hunk 4]
```diff
@@ -87,13 +97,24 @@ void Simulator::clearInstances()
 	}
 }
 
-void Simulator::connect()
+bool Simulator::connect()
```
Comment (bradleynull) [inline monodrive/core/src/Simulator.cpp:100]: This `bool` isn't checked anywhere. Can we add this the output of `getInstance()` and return the simulator object through a parameter?

File: monodrive/core/src/Simulator.cpp [hunk 5]
```diff
@@ -160,6 +184,7 @@ bool Simulator::sendCommand(ApiMessage message, nlohmann::json *response)
 	{
 		return true;
 	}
+	std::cerr << res.get_message().dump() << std::endl;
```
Comment (bradleynull) [inline monodrive/core/src/Simulator.cpp:187]: Remove debug output.

File: monodrive/core/src/Simulator.h [hunk 6]
```diff
@@ -40,7 +40,11 @@ class Simulator
 	}
 	bool stateStepSampleAll(std::vector<std::shared_ptr<Sensor>>& sensors, const nlohmann::json& state);
 	void stepSampleAll(std::vector<std::shared_ptr<Sensor>>& sensors, int stepIndex, int numSteps);
-	void sampleAll(std::vector<std::shared_ptr<Sensor>>& sensors);
+	// samples all sensors on the server (even those not in the list)
+	// faster but make sure you know what you're doing
+	bool sampleAll(std::vector<std::shared_ptr<Sensor>>& sensors);
```
Comment (bradleynull) [inline monodrive/core/src/Simulator.h:45]: Maybe a little more documentation here. What are the cases where this should not be used?

File: monodrive/core/src/SimulatorCommands.h [hunk 7]
```diff
@@ -54,6 +54,7 @@
  */
 #define EgoControl_ID				"EgoControl_ID" 
 #define SampleSensorsCommand_ID "SampleSensorsCommand_ID"
+#define SampleSensorCommand_ID "SampleSensorsCommand_ID"
```
Comment (bradleynull) [inline monodrive/core/src/SimulatorCommands.h:57]: `SampleSensorCommand_ID` and `SampleSensorsCommand_ID` both map to `"SampleSensorsCommand_ID"`. Is this intentional?

File: examples/cpp/lane_follower/fixed_step_lidar.cpp (hunk 1 above)
Comment (stephendb) [reply examples/cpp/lane_follower/fixed_step_lidar.cpp]: thanks

File: examples/cpp/multi_process_example/simulator_controller.cpp (hunk 2 above)
Comment (stephendb) [reply examples/cpp/multi_process_example/simulator_controller.cpp:31]: fixed

File: monodrive/core/src/Simulator.cpp (hunk 3 above)
Comment (stephendb) [reply monodrive/core/src/Simulator.cpp]: clean

File: monodrive/core/src/Simulator.cpp (hunk 4 above)
Comment (stephendb) [reply monodrive/core/src/Simulator.cpp:99]: just started the process, more things like that should either throw or be a bool, will look at stuff like this in the next refactor/hardening

File: monodrive/core/src/Simulator.cpp (hunk 5 above)
Comment (stephendb) [reply monodrive/core/src/Simulator.cpp:186]: not debug, that is the message in the failure case, apparently we used to just dump the message that the server sent

File: monodrive/core/src/Simulator.h (hunk 6 above)
Comment (stephendb) [reply monodrive/core/src/Simulator.h:47]: added

File: monodrive/core/src/SimulatorCommands.h (hunk 7 above)
Comment (stephendb) [reply monodrive/core/src/SimulatorCommands.h]: fixed, now is SampleSensorListCommand_ID

File: monodrive/core/src/Simulator.cpp [hunk 8]
```diff
@@ -212,13 +236,44 @@ void Simulator::stepSampleAll(std::vector<std::shared_ptr<Sensor>>& sensors, int
...
+		WaitForSamples(sensors);
+	}
+	else{
+		std::cerr << "Failed to sample sensors." << std::endl;
+		return false;
+	}
+	return true;
+}
+
+void Simulator::WaitForSamples(const std::vector<std::shared_ptr<Sensor>>& sensors)
```
Comment (devinaconley) [inline monodrive/core/src/Simulator.cpp:275]: use lowerCamelCase in client

File: monodrive/core/src/Simulator.cpp [hunk 9]
```diff
@@ -107,38 +127,41 @@ void Simulator::stop()
...
+				std::cout << sendCommand(ApiMessage(1001, REPLAY_ConfigureTrajectoryCommand_ID, true, config.scenario)) << std::endl;
+			}
 		}
 	}
 
-	if (!config.weather.empty())
+
+	if (config.weather.empty())
+	{
+		std::cout << "Skipping Weather Config, no weather config set." << std::endl;
```
Comment (devinaconley) [inline monodrive/core/src/Simulator.cpp:162]: I think we need to add a `verbose=` option in the simulator object. Can do this in a later PR

**Pull request merged**